### ISSUE TITLE
[RML]_DT_RF.R

### ISSUE THREAD

**Issue by takytaky** [takytaky/rusita#6]
# 작업 디렉토리 설정

getwd()
setwd("~/[006]_DT_RF")
getwd()

# Wine Data ####

# Train_Data - quality 포함(3 ~ 9)

Train_Data <- read.csv("data/Train_Data.csv")

str(Train_Data)

table(Train_Data$quality)

barplot(table(Train_Data$quality), 
        col = rainbow(7))

# Test_Data - quality 미포함(NA)

Test_Data <- read.csv("data/Test_Data.csv")

str(Test_Data)

Test_Data$quality

# Model Validation Review ####
# quality Regression Model

Model_lm <- lm(quality ~ ., data = Train_Data)

# 생성된 모델로 Test_Data에 적용

Pred_Quality <- predict(Model_lm, Test_Data)

round(head(Pred_Quality))

Train_Data$quality[1:6]

# 예측 결과를 파일로 저장
# Model Validation 과정 필요

write.csv(Pred_Quality, "data2/PreD.csv")

# Validation Approach
# Train_Data:Validation_Data 분리(7:3)

set.seed(8888)

# 70% : Training Data

TD_7 <- sample(1:nrow(Train_Data),
               nrow(Train_Data) * 0.7)

# Training Data로 Modeling

Model_lm <- lm(quality ~ ., 
               data = Train_Data[TD_7,])

# 30% : Validation Data에 모델 적용

Pred_Qual_lm <- predict(Model_lm, 
                        Train_Data[-TD_7,])

# Output Data(교사데이터)

Valid_Data <- Train_Data$quality[-TD_7]

# 오차계산

mean((Valid_Data - Pred_Qual_lm)^2)

table(Valid_Data)
table(Valid_Data, round(Pred_Qual_lm),
      dnn = c("Actual", "Predicted"))

# 다양한 모델을 생성하여 비교
# Model_j <- lm(quality ~ . PH * density, data = Train_Data[TD_7,])
# Model_k <- lm(quality ~ . + poly(PH, 3), data = Train_Data[TD_7,])

# Decision Tree

# Orange Data ####
# install.packages("ggplot2")

library(ggplot2)

# Orange Data File 읽어오기

Orange <- read.csv("data/Orange.csv")

str(Orange)

head(Orange)

# 탐색적 데이터 분석

# weight ~ model

ggplot(Orange, aes(x = model, y = weight)) +
  geom_boxplot(aes(fill = model),
               outlier.color = "red",
               outlier.shape = "*",
               outlier.size = 20)

# sugar ~ model

ggplot(Orange, aes(x = model, y = sugar)) +
  geom_boxplot(aes(fill = model),
               outlier.color = "red",
               outlier.shape = "*",
               outlier.size = 20)

# acid ~ model

ggplot(Orange, aes(x = model, y = acid)) +
  geom_boxplot(aes(fill = model),
               outlier.color = "red",
               outlier.shape = "*",
               outlier.size = 20)

# color ~ model

ggplot(Orange, aes(factor(color))) +
  geom_bar(aes(fill = model))

# model

ggplot(Orange, aes(model)) + 
  geom_bar(aes(fill = model))

# Train:Test - (8:2)

set.seed(7777)

# 80% : Training Data Index

TD_8 <- sample(1:nrow(Orange),
               nrow(Orange) * 0.8)

# 20% : Vaidation DataFrame

Valid_Orange <- Orange[-TD_8, ]

# tree Package - tree() 
# install.packages("tree")

library(tree)

# Modeling - tree()
# Entropy Index : 데이터 집합의 혼잡도 기준
# 이진 재귀 분할법(Binary Recusive Partitioning)

Model_dt <- tree(model ~ ., 
               data = Orange[TD_8, ]) 

Pred_Qual_dt <- predict(Model_dt, 
                        Orange[-TD_8,])

structure(Model_dt)

summary(Model_dt)

plot(Model_dt)
text(Model_dt)

# rpaty Package - rpart()
# CART(Classification And Regression Tree)
# install.packages("rpart")
# install.packages("rpart.plot")

library(rpart)
library(rpart.plot)

# Modeling - rpart()
# minsplit : 분류 시 포함할 최소 데이터의 개수
# 분순도(Impurity) 기반의 분리(splitting) 실행

Model_RP <- rpart(model ~ .,
                  data = Orange[TD_8,], 
                  control = rpart.control(minsplit = 2))

# Model Structure

structure(Model_RP)

# Model Summary

summary(Model_RP)

# Model Visualization
# 범례, 실제 Train Data 비율, 조건식 포함

rpart.plot(Model_RP)

# 20% : Validation Data - Classification
# type = "class" - 분류 결과를 레이블로 출력
# type = "prob"  - 분류 결과를 확률로 출력

Pred_Qual_RP <- predict(Model_RP, 
                        Orange[-TD_8,], 
                        type = "class")

# Performance Check
# caret Package - confusionMatrix()
# Accuracy, Cohen's Kappa
# Precision, Recall(Sensitivity), Specificity
# Accuracy    : 전체 정확도
# Kappa       : 평가자간 일치도(0 ~ 1 사이값)
# Precision   : T로 예측한것 중 실제 T의 비율
# Recall      : 실제 T 중 모델이 T로 맞춘 비율
# Specificity : 실제 F 중 모델이 F로 맞춘 비율

# install.packages("caret")

library(caret)

confusionMatrix(data = Pred_Qual_RP, 
                reference = Valid_Orange$model)

# CP(Complexity Parameter)
# 단순한모델 : 직관적이지만 성능이 낮음
# 복잡한모델 : 이해가 어렵고 계산 시간이 길어짐
# 훈련데이터에 최적화(과적합)되는 문제 발생
# CP : 노드를 분리하는데 필요한 복잡성(Complexity/Cost)
# 최대CP값 도달 시 더 이상 분류작업을 진행하지 않음
# 최적을 CP 설정으로 범용적 모델 생성 가능

Model_RP$cptable

printcp(Model_RP)

plotcp(Model_RP)

# Pruning ####

# Part-1

# CP값을 0.13으로 지정

Model_RP_P <- prune(Model_RP,
                    cp = 0.13)

rpart.plot(Model_RP_P)

Pred_Qual_RP_P <- predict(Model_RP_P, 
                          Orange[-TD_8,], 
                          type = "class")

confusionMatrix(data = Pred_Qual_RP_P, 
                reference = Valid_Orange$model)

printcp(Model_RP_P)

plotcp(Model_RP_P)

# Part-2

# tree() 함수에 적용

plot(Model_dt)
text(Model_dt)

# k-fold Cross-Validation

Model_CVT <- cv.tree(Model_dt, FUN = prune.misclass)
plot(Model_CVT)

# 3개의 가지로 가지치기(Pruning)

PT1 <- prune.misclass(Model_dt, best = 3)

plot(PT1)
text(PT1)

# 개별실습-1 ####
# Heart.csv - AHD(심장병 유무 정보)
# Train:Test(7:3)

HEART <- read.csv("data/Heart.csv")

str(HEART)

head(HEART)

table(HEART$AHD)
  

# Conditional Inference Tree
# 순열 검정 기반 비편향 재귀 분할법
# Unbiased Recursive Partitioning based on Permutation Test
# p검정 기준으로 가지치기 변수를 결정
# 편향의 위험이 적음(단, 입력변수가 31개로 제한됨)

# party Package - ctree()
# install.packages("party")

library(party)

# iris Data

iris <- read.csv("data/IRIS.csv")

str(iris)

head(iris)

iris <- iris[, -1]

head(iris)

# Train Data:Validation Data(7:3)

set.seed(123)

IDX <- sample(2, size = nrow(iris),
              replace = TRUE,
              prob = c(0.7, 0.3))

IDX

iris_TD <- iris[IDX == 1, ]
str(iris_TD)

iris_VD <- iris[IDX == 2, ]
str(iris_VD)

# Modeling

Model_ct <- ctree(Species ~ ., data = iris_TD)

structure(Model_ct)

summary(Model_ct)

# Visualization

plot(Model_ct, type = "simple")

plot(Model_ct)

# Validation Data에 모델 적용

Pred_Qual_ct <- predict(Model_ct, 
                        newdata = iris_VD)

table(Pred_Qual_ct)

# Model Performance

confusionMatrix(Pred_Qual_ct, iris_VD$Species)

# Random Forest - Ensemble ####

# 의사결정나무의 앙상블 모델
# 여러개의 나무를 학습 후 종합하는 모델
# 이상치나 편향된 분포에 민감하지 않음

# Bootstrap:단순 복원 임의추출법으로 추출한
#           크기가 동일한 여러개의 표본 데이터
#           학습데이터를 늘리는 효과 발생
# Bagging:여러개의 Bootstrap 데이터를 생성하고
#         Bootstrap된 각 데이터를 모델링한 후
#         최종적인 예측모델을 생성
#         통계적학습방법의 분산을 줄이는 과정

# install.packages("randomForest")

library(randomForest)

# iris Data(7:3)

str(iris_TD)

str(iris_VD)

# Modeling
# Number of trees: 500
# No. of variables tried at each split: 2

Model_En <- randomForest(Species ~ .,
                         data = iris_TD,
                         importance = TRUE)

structure(Model_En)

summary(Model_En)

# 각 변수의 중요도(importance = TRUE)
# MeanDecreaseAccuracy:분류정확도 개선에 기여한 변수의 수치
# MeanDecreaseGini:노드 불순도(불확실성) 개선에 기여한 변수의 수치
# Petal.Length가 가장 크게 기여

Model_En$importance

varImpPlot(Model_En)

# Validation Data에 모델 적용

Pred_Qual_En <- predict(Model_En, 
                        newdata = iris_VD)

table(Pred_Qual_En)

# Model Performance

confusionMatrix(Pred_Qual_En, iris_VD$Species)

# Wine Data
# 500개의 Tree 생성(Bagging)
# Number of trees: 500
# No. of variables tried at each split: 4

Model_rf <- randomForest(quality ~ ., 
                         data = Train_Data[TD_7,],
                         importance = TRUE)

structure(Model_rf)

summary(Model_rf)

Model_rf$importance

varImpPlot(Model_rf)

# 30% : Validation Data에 모델 적용

Pred_Qual_rf <- predict(Model_rf, 
                        Train_Data[-TD_7,])

# 오차계산

mean((Valid_Data - Pred_Qual_rf)^2)

# Hyper_Parameter ####
# 학습전에 지정하는 파라미터
# summary(Model_rf)
# ?randomForest
# ntree:추출되는 나무의 개수
# mtry :적용되는 변수의 개수
# nodesize:마지막 노드에 지정된 데이터 개수

# mtry = 1

Model_rf <- randomForest(formula = quality ~ .,
                         data = Train_Data[TD_7,],
                         mtry = 1)

Pred_Qual_rf <- predict(Model_rf, Train_Data[-TD_7,])

mean((Valid_Data - Pred_Qual_rf)^2)

# mtry = 3

Model_rf <- randomForest(formula = quality ~ .,
                         data = Train_Data[TD_7,],
                         mtry = 3)

Pred_Qual_rf <- predict(Model_rf, Train_Data[-TD_7,])

mean((Valid_Data - Pred_Qual_rf)^2)

# mtry = 6

Model_rf <- randomForest(formula = quality ~ .,
                         data = Train_Data[TD_7,],
                         mtry = 6)

Pred_Qual_rf <- predict(Model_rf, Train_Data[-TD_7,])

mean((Valid_Data - Pred_Qual_rf)^2)

# mtry = 9

Model_rf <- randomForest(formula = quality ~ .,
                         data = Train_Data[TD_7,],
                         mtry = 9)

Pred_Qual_rf <- predict(Model_rf, Train_Data[-TD_7,])

mean((Valid_Data - Pred_Qual_rf)^2)

# mtry = 12(Bagging)

Model_rf <- randomForest(formula = quality ~ .,
                         data = Train_Data[TD_7,],
                         mtry = 12)

Pred_Qual_rf <- predict(Model_rf, Train_Data[-TD_7,])

mean((Valid_Data - Pred_Qual_rf)^2)

# ranger Package ####
# install.packages("ranger")

library(ranger)

Model_rg <- ranger(quality ~ ., 
                   data = Train_Data[TD_7,])

Pred_Qual_rg <- predict(Model_rg, 
                        Train_Data[-TD_7,])$predictions

mean((Valid_Data - Pred_Qual_rg)^2)

# mtry = 3

Model_rg <- ranger(formula = quality ~ .,
                   data = Train_Data[TD_7,],
                   mtry = 3)

Pred_Qual_rg <- predict(Model_rg, 
                        Train_Data[-TD_7,])$predictions

mean((Valid_Data - Pred_Qual_rg)^2)

# mtry=3, num.trees=1000, min.node.size = 3

Model_rg <- ranger(formula = quality ~ .,
                   data = Train_Data[TD_7,],
                   mtry = 3,
                   num.trees = 1000,
                   min.node.size = 3)

Pred_Qual_rg <- predict(Model_rg, 
                        Train_Data[-TD_7,])$predictions

mean((Valid_Data - Pred_Qual_rg)^2)

# Boosting Model ####

# Bagging :선택된 Tree간에 관계가 없음

# Boosting:이전 모델이 다음 모델에 영향을 줌
#          순차적(Sequential) 방법
#          이전 분류된 학습결과로 
#          다음 학습데이터의 샘플가중치를 조정

# Gradient Boosting Model
# install.packages("gbm")
# Hyper_Parameter
# n.trees:생성되는 나무의 개수
# shrinkage:Learning Rate

library(gbm)

# Validation Approach

Model_gbm <- gbm(formula = quality ~ .,
                 data = Train_Data[TD_7,],
                 n.trees = 4000,
                 shrinkage = 0.1)

structure(Model_gbm)

summary(Model_gbm)

Pred_Qual_gbm <- predict(Model_gbm, 
                         Train_Data[-TD_7,],
                         n.trees = 1000)

mean((Valid_Data - Pred_Qual_gbm)^2)

# Cross Validation

Model_gbm <- gbm(formula = quality ~ .,
                 data = Train_Data,
                 cv.folds = 5,
                 n.trees = 10000,
                 shrinkage = 1)

structure(Model_gbm)

summary(Model_gbm)

Pred_Qual_gbm <- predict(Model_gbm, 
                         Train_Data[-TD_7,],
                         n.trees = 1500)

mean((Valid_Data - Pred_Qual_gbm)^2)

# 개별실습-2 ####
# Credit.csv - default(연체유무)
# Train:Test(7:3)

CRDT <- read.csv("data/Credit.csv")

str(CRDT)

head(CRDT)

table(CRDT$default)
